### PR TITLE
Modernize Xpect

### LINKED ISSUE #320: Modernize Xpect
Xpect still uses the old xtext workflow (see what new Xtext project wizard generates) and the old xtext generator.(xpand based)
this is cause of
https://github.com/eclipse/Xpect/blob/018800b3af074c402b2aa53174e2457ea6f48046/org.eclipse.xpect/src/org/eclipse/xpect/GenerateXpect.mwe2#L124C6-L124C6
which uses
https://github.com/meysholdt/xtext_stateful_lexer/blob/48a4201ea3e0e52743af553ffaea89ba2d79144c/de.itemis.statefullexer/src/de/itemis/statefullexer/StatefulLexerFragment.xtend#L22

- [x] this needs to be rewritten on base of
https://github.com/eclipse/xtext/blob/main/org.eclipse.xtext.xtext.generator/src/org/eclipse/xtext/xtext/generator/parser/antlr/ex/ExternalAntlrLexerFragment.java or by an adjusted version of that

we should move the new fragment to a new plugin in the xpect project.
we also need to make sure content assist and highlighting lexer get put to the ide project.

once this is done we can move in xpect

- [x] to a newer minimal Xtext version (ExternalAntlrLexerFragment above requires at lease 2.20), i propose a much newer version
- [x] the new xtext workflow (also create an ide project), compare generated code, specially the runtime and ui moduels for diffs
- [x] thus get rid of xpand and dependency to xtexts for removal old generator
- [x] then we can change the pom to get rid of the fornax plugin to execute the workflow which is also no longer maintained (see also what new Xtext project wizard generates)
- [x] we can and should also use newer tycho version (and get rid of outdated tycho 1.7/2.x stuff)
- [x] and with that have better separation between compile time and runtime deps (put dep. to xtext.xtext.generator to build.properties, e.g. as new xtext project wizard generates it)
- [x] simplifiy target file setup in tycho as in xtext
- [x] Optional: repair oomph
- [x] Check if we can fix some Xtext and Java deprecations

also the workflows in the examples need to be modernized, but that should be straight forward.
we also should get of deprecated xtext junit4 there

maybe we also should create a 0.2.0 branch with the current state and bump master to 0.3.0 then

in an alternative approach we would update target platform, xtext+java version, workflow, tycho and pom first and use fragment adapter (see history of /org.eclipse.xtext.ui.codetemplates/src/org/eclipse/xtext/ui/codetemplates/GenerateCodetemplateLanguages.mwe2) to call old workflow and then create the new fragment.
in this case we could make sure the "new dependencies" work first before introducing the changed code.

regardings versions:
- Java 11 + Eclipse 2022-03 + Xtext 2.31
- Java 17 + Eclipse 2023-06 + Xtext 2.31 (or 2.32 when it is already out) (as we will have to drop 11 Support in Xtext anyway one day i propose to go this way)

### PR DESCRIPTION
**Pull request by tjeske**
Resolves #320

### CODE REVIEW

**Comment by cdietrich** [comment]
@szarnekow would be nice if you also could have a look

File: Jenkinsfile [hunk 1]
```diff
@@ -40,32 +39,16 @@ timestamps() {
...
-                        sh "${mvnHome}/bin/mvn -P!plugins -P!xtext-examples -Dtarget-platform=eclipse_4_5_0-xtext_2_14_0 ${mvnParams} clean integration-test"
-                    }finally {
-                        junit '**/TEST-*.xml'
-                    }
-                }
-
-                stage('test with Eclipse 2022-03 and Xtext nighly') {
-                    try{
-                        sh "JAVA_HOME=${java11Home} ${mvnHome}/bin/mvn -P!plugins -P!xtext-examples -Dtycho-version=2.7.5 -Dtarget-platform=eclipse_2022_03-xtext_nightly ${mvnParams} clean integration-test"
+                        sh "JAVA_HOME=${java11Home} ${mvnHome}/bin/mvn -P!plugins -P!xtext-examples -Dtycho-version=2.7.5 -Dtarget-platform=eclipse_2023_06-xtext_nightly ${mvnParams} clean integration-test"
```
Comment (cdietrich) [inline Jenkinsfile:51]: this does not seem to work.
202306 will need java 17 anyway

**Comment by cdietrich** [comment]
TODO / pot. follow ups

- Deprecated XMLReaderFactory
- unused imports
- set minimal java version to 17
- use mwe 2.15 release in nightly target 
- use 2023-09 in nightly target 
- use newer orbit (check xtext targets)

**Comment by cdietrich** [comment]
```
org.eclipse.xpect/META-INF/MANIFEST.MF:Require-Bundle: org.eclipse.xpect.mwe2;bundle-version="0.2.0",
org.eclipse.xtext.example.domainmodel.xpect.tests/META-INF/MANIFEST.MF: org.eclipse.xpect.xtext.lib;bundle-version="0.2.0",
org.eclipse.xtext.example.domainmodel.xpect.tests/META-INF/MANIFEST.MF: org.eclipse.xpect.xtext.xbase.lib;bundle-version="0.2.0",
```
might want to bump versions here too

File: Jenkinsfile (hunk 1 above)
Comment (tjeske) [reply Jenkinsfile:51]: Oh yes, this cannot work. Can we test the Jenkinsfile of this PR (after I have fixed it)? Jenkins builds this PR with the Jenkinsfile from master which fails (due to Java 8 etc.).

**Comment by tjeske** [comment]
> nice change. great achievement
> 
> TODO / pot. follow ups
> 
>     * Deprecated XMLReaderFactory
> 
>     * unused imports
> 
>     * set minimal java version to 17
> 
>     * use mwe 2.15 release in nightly target
> 
>     * use 2023-09 in nightly target
> 
>     * use newer orbit (check xtext targets)

Thanks for having a look!

I didn't want to replace the Java `XMLReaderFactory` stuff in this PR, but after a closer look I think it is not a big change. I will give it a try tomorrow.

Okay, if nobody is on Java 11 anymore, I set minimum Java version to 17 and switch to latest Eclipse release.

I will refresh the deps etc. tomorrow.

**Comment by cdietrich** [comment]
We can do the Java 17 move to a second step / pr

**Comment by kuniss** [comment]
> Okay, if nobody is on Java 11 anymore, I set minimum Java version to 17 and switch to latest Eclipse release.

Is there really a need to switch to Java 17? As long as Java 11 LTS is running we may stay with it? I would switch to Java 17 when Eclipse is requiring Java 17...

**Comment by kuniss** [comment]
> > > Okay, if nobody is on Java 11 anymore, I set minimum Java version to 17 and switch to latest Eclipse release.
> > 
> > 
> > Is there really a need to switch to Java 17? As long as Java 11 LTS is running we may stay with it? I would switch to Java 17 when Eclipse is requiring Java 17...
> 
> Eclipse 2023-06 requires Java >= 17.

Guess, I'm out of date... :smile:

**Comment by cdietrich** [comment]
changes look good so far, but i still want to set this up locally to see how workspace looks like.
anyway i propose to bring a first version to the mondernize branch so this can be tested more broadly

**Pull request merged**